### PR TITLE
Redo roles and permissions

### PR DESCRIPTION
**Pull request by pstaabp**
This branch does a number of things with roles and permissions: 
- all roles and permissions are defined in `permissions.dist.yml`
- admins can copy this file over and change permissions/add roles, etc. 
- there is a role called `course_admin`, which can have additional permissions that instructors don't have. 
- there are both server-side permissions and ui permissions.
- there is a script `bin/update_perms.pl` which loads both the roles and permissions in the `permissions.yml` file into the database.
- on the server side, the permissions are checked by querying the database. 
- on the UI side, after login, the permissions are loaded in the store to be able to be queried whenever. 
- every route is now checked that the user has permission to visit that route. 

Some things that still need to be addressed:
- I branched this off main and there are some more tests in `t/mojolicious` that are in #88 and #89 that will need to be checked. 
- Wondering if this is sufficient or do we need another level of permissions similar to ww2 that is more fine grained?  If so, I think this can be added in a similar manner. 
- There's still some issues during testing, so if a user doesn't have permission, an error is logged, the user is not prevented from going to that page.  Clearly, this needs to be updated before being merged. 
- The routes on the ui-side are crude regexps and we may want to put the regexp in the permissions file.

### CODE REVIEW

**Comment by drdrew42** [comment]
On first glance, I'm not a fan of swapping out the UserRole.unknown for empty string. If we're ENUM-ing the roles as we are, there needs to be a role for those who aren't logged in. Maybe change 'unknown' to 'notloggedin' or something indicative of the situation it is meant to represent. This serves to make the code more readable.

I'm also not clear on the new role: 'course_admin' ... what is this for? What will this provide for that the admin flag on the User doesn't already provide?

Also, it looks like you're building a many-to-many relationship (roles <-> actions) through a JSON field that holds an array. This should be a mapping table instead.

**Comment by pstaabp** [comment]
This is ready to be tested now.  
* rebuild database (important as the roles/permissions are now stored in the database)
* run perl tests
* run typescript tests
* test UI (no new functionality, but some needed changes for permissions)

**Comment by drdrew42** [comment]
Visit `/webwork3` without a cookie, view console:
* `[routing] The user does not have permission to visit '/login'`
* `[Vue Router warn]: The "next" callback was called more than once in one navigation guard when going from "/" to "/login". It should be called exactly one time in each navigation guard. This will fail in production.`

Log in as lisa, then logout, view console:
* `[Vue Router warn]: The "next" callback was called more than once in one navigation guard when going from "/users/3/courses" to "/login". It should be called exactly one time in each navigation guard. This will fail in production.`

Something is not being handled correctly in `index.ts` where the permission store is being put to use... it's working in dev mode, but these warnings need to be heeded -- this won't work in production.

**Comment by pstaabp** [comment]
I merged this in.  I still get errors if not authenticated, but not the call to next() twice that you saw.  I think we discussed this earlier about more elegantly handling session expiring.  Was this supposed to handle going to /login?

**Comment by drdrew42** [comment]
It certainly handles redirect to /login when logging out (which wasn't happening before -- we'd get the url for /login, but the page wasn't loading the login vue). It also handles redirection to the user_courses page when attempting to access any url for which the user does not have permission. I think we can extend this change to cover cookie timeout as well...

This does *not* address attempting to visit a url, being redirected to login, and afterwards maintaining the initial target url.

**Comment by drgrice1** [comment]
@pstaabp: Sorry I didn't get to this today.  I should have time tomorrow though.

File: lib/DB/Schema/Result/DBPermission.pm [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	}
+);
+
+__PACKAGE__->inflate_column(
+	'admin_required',
+	{
+		inflate => sub { return shift ? Mojo::JSON->true : Mojo::JSON->false; },
+		deflate => sub { return shift; }
+	}
+);
```
Comment (drgrice1) [inline lib/DB/Schema/Result/DBPermission.pm:74]: It would be good to create an InflateColumn::Serializer class for these boolean columns instead of repeating this for every boolean column.  This is where the technique you used for the JSON columns would come in handy.

File: lib/DB/Schema/Result/User.pm [hunk 2]
```diff
@@ -124,7 +124,7 @@ __PACKAGE__->inflate_column(
 __PACKAGE__->set_primary_key('user_id');
 __PACKAGE__->add_unique_constraint([qw/username/]);
 
-__PACKAGE__->has_many(course_users => 'DB::Schema::Result::CourseUser', { 'foreign.user_id' => 'self.user_id' });
+__PACKAGE__->has_many(course_users => 'DB::Schema::Result::CourseUser', 'user_id');
```
Comment (drgrice1) [inline lib/DB/Schema/Result/User.pm:127]: Why the change here?  These are equivalent, and what was wrong with the shorthand?

File: lib/DB/Schema/ResultSet/User.pm [hunk 3]
```diff
@@ -272,9 +271,14 @@ or an arrayref of C<DBIx::Class::ResultSet::User>
...
-		'course_id' => $course->course_id
-	});
+	my @course_users = $self->rs('CourseUser')->search(
+		{
+			'course_id' => $course->course_id
+		},
+		{
+			prefetch => [qw/role/]
+		}
+	);
```
Comment (drgrice1) [inline lib/DB/Schema/ResultSet/User.pm:281]: Why is this so spread out?  It can be done in two lines if you change it to
```suggestion
	my @course_users =
		$self->rs('CourseUser')->search({ 'course_id' => $course->course_id }, { prefetch => [qw/role/] });

```
pertidy accepts that.  The same can be done on lines 337 - 344 below.

File: lib/WeBWorK3/Controller/Permission.pm [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+sub checkPermission ($c) {
+	# seems like $c->current_user doesn't have access to the course_id, so we never
+	# get a CourseUser.
+	my $user = $c->current_user;
+
+	# The controller/action pair is stored in the $c->match->stack array at the end.
+	# See https://docs.mojolicious.org/Mojolicious/Guides/Routing#Under
+	my $controller = $c->match->stack->[1]->{controller};
+	my $action     = $c->match->stack->[1]->{action};
```
Comment (drgrice1) [inline lib/WeBWorK3/Controller/Permission.pm:44]: There is unnecessary de-referencing here.  This should be
```suggestion
	my $controller = $c->match->stack->[1]{controller};
	my $action     = $c->match->stack->[1]{action};
```
Even the Mojolicious documentation you referenced does it this way.

There is more of this below in this file (and probably elsewhere).

@dpvc pointed this sort of thing out to me a few years ago.  webwork2 has massive amounts of this sort of unnecessary de-referencing.

File: lib/WeBWorK3/Controller/Problem.pm [hunk 5]
```diff
@@ -4,87 +4,110 @@ use strict;
 
 use Mojo::Base 'Mojolicious::Controller', -signatures;
 
-sub addProblem ($self) {
-	my $problem = $self->schema->resultset('SetProblem')->addSetProblem(
-		params => {
-			course_id => int($self->param('course_id')),
-			set_id    => int($self->param('set_id')),
-			%{ $self->req->json }
+sub getAllProblems ($c) {
```
Comment (drgrice1) [inline lib/WeBWorK3/Controller/Problem.pm:7]: This conversion of `$self` to `$c` and `$self` to `$app` is exactly the sort of thing that you should do in a dedicated pull request.  If that is all that is in the pull request, it is easy to review and would be merged quickly.  By doing this in this pull request you make it so much harder to review, and slow down the merge of this pull request.

**Review by drgrice1 — COMMENTED** [review]
This is just a start of my review.  I am out of time for now, and will continue it tomorrow.

File: lib/DB/Schema/Result/User.pm (hunk 2 above)
Comment (drdrew42) [reply lib/DB/Schema/Result/User.pm:127]: isn't the changed version technically the 'shorthand'? When we have the same id on two tables, this is more concise?

Comment (drgrice1) [reply lib/DB/Schema/Result/User.pm:127]: Hah.  It seems that I read the diff in reverse.  The changed version is the shorthand.  Never mind then.

File: lib/WeBWorK3/Controller/Problem.pm (hunk 5 above)
Comment (drdrew42) [reply lib/WeBWorK3/Controller/Problem.pm:7]: Yes, absolutely. But, to be clear, this is not a request to revert -- just a note for the future? These changes should be one cleanup PR, separate from a feature PR, yeah?

File: lib/WeBWorK3/Controller/Permission.pm (hunk 4 above)
Comment (drdrew42) [reply lib/WeBWorK3/Controller/Permission.pm:44]: @pstaabp and I talked about this -- yes, we have absolutely picked up some bad habits from the ww2 codebase.

File: t/mojolicious/003_users.t [hunk 6]
```diff
@@ -72,14 +75,37 @@ my $added_user_to_course = {
...
+# Check if a non-existent user is a global user
+$t->get_ok('/webwork3/api/courses/1/users/non_existent_user/exists')->status_is(200)
+	->content_type_is('application/json;charset=UTF-8');
+
+is_deeply($t->tx->res->json, {}, 'checkUserExists: check that a non-existent user returns {}');
+
+# Testing that booleans returned from the server are JSON booleans.
+# the first user is the admin
+
+my $admin_user = firstval { $_->{username} eq 'admin' } @all_users;
```
Comment (drgrice1) [inline t/mojolicious/003_users.t:95]: Something for the future:
I don't really like the usage of the `List::MoreUtils` module for only the `firstval` method as you have done all over in the tests and this is the only way this module is ever used.  That is something that can be done quite easily with native perl.  Just change this to
```suggestion
my $admin_user = ({ $_->{username} eq 'admin' } @all_users)[0];
```
That code is equivalent.  It returns the first element of the array, and is undefined if the array is empty.

If this is changed, that is one less module that we need as a dependency.

File: tests/unit-tests/utils.spec.ts [hunk 7]
```diff
@@ -2,82 +2,84 @@
 
 import { pick, mapValues, invert, pickBy, random } from 'src/common/utils';
 
-test('testing pick', () => {
-	const obj = { a: 1, b: 2, c: 3, d: 4 };
+describe('Testing all utility functions', () => {
```
Comment (drgrice1) [inline tests/unit-tests/utils.spec.ts:5]: For the future again:
Surely you see that this change of adding descriptions is exactly the sort of completely unrelated change that should be done in another pull request.  These changes would even have almost no conflict with the other changes in this pull request, and if there are any minor conflicts they could easily be resolved.  This is what branches are for in git.

**Review by drgrice1 — CHANGES_REQUESTED** [review]
This finishes my review of this pull request.  See my comments in the code for some things that should be changed, and some other things to think about for the future.

And the big show stopper ... running the quasar dev server and morbo, and opening localhost:8080 in the browser results in a blank screen, and webwork3 never loads.  There is the error "[error] [routing] User #0 does not have the permission to visit /login" in the console.  The user number can be different.  When testing in Firefox I got #0, and in Chrome I got #3.  However in both browsers the UI does not load.  In Firefox the browser hangs even, and it takes a while for it to finish loading.  This may be because I have the Vue dev tools installed there, but not in Chrome.

File: src/router/index.ts [hunk 8]
```diff
@@ -28,5 +31,20 @@ export default route(() => {
 		),
 	});
 
-	return Router;
+	router.beforeResolve((to) => {
+		// must be logged in, or else headed to the login page
+		const session_store = useSessionStore();
+		if (!session_store.logged_in && to.name !== 'login') return { name: 'login' };
+
+		// otherwise, we're logged in, so check permissions on the route
```
[the remark this answers is not in the document]
Comment (drdrew42) [reply src/router/index.ts:39]: This one is on me, I think. This all gets ripped out in an upcoming PR to handle session/cookie timeout. I'll work on eliminating the royal 'we' 👑  in the future.

File: src/stores/users.ts [hunk 9]
```diff
@@ -135,7 +135,7 @@ export const useUserStore = defineStore('user', {
 		 * Updates the given user in the database and in the store.
 		 */
 		async updateUser(user: User): Promise<User | undefined> {
-			if (!user.isValid()) return invalidError(user, 'The updated user is invalid');
+			if (!user.isValid()) await invalidError(user, 'The updated user is invalid');
```
[the remark this answers is not in the document]
Comment (drdrew42) [reply src/stores/users.ts:138]: Yes. I'm guessing this is stemming from confusion over promises? `updateUser` intends to return a promise. On an error, however, it seems that `invalidError` is here to put a promise wrapper around the Error. Then, the `updateUser` caller expects a promise and should already be `await`ing (so it's incorrect to await here...)

File: src/layouts/MainLayout.vue [hunk 10]
```diff
@@ -35,10 +35,11 @@ import MenuBar from './MenuBar.vue';
 
 import { instructor_views, admin_views, ViewInfo, student_views } from 'src/common/views';
 import { useRoute } from 'vue-router';
+import { logger } from 'boot/logger';
+
```
[the remark this answers is not in the document]
Comment (pstaabp) [reply src/layouts/MainLayout.vue:39]: I'll be more careful with this. This looks like I just did some reorganization, but clearly, it doesn't belong here.

File: t/db/build_db.pl [hunk 11]
```diff
@@ -265,7 +280,12 @@ sub addUserSets {
 
 sub addProblemPools {
 	say 'adding problem pools' if $verbose;
-	my @problem_pools = loadCSV("$main::ww3_dir/t/db/sample_data/pool_problems.csv");
+	my @problem_pools = my @problem_pools_from_file = loadCSV(
+		"$main::ww3_dir/t/db/sample_data/pool_problems.csv",
+		{
+			non_neg_int_fields => ['library_id']
+		}
+	);
```
[the remark this answers is not in the document]
Comment (pstaabp) [reply t/db/build_db.pl:288]: I'll move over to this style.

File: t/mojolicious/003_users.t (hunk 6 above)
Comment (pstaabp) [reply t/mojolicious/003_users.t:95]: Thanks for this.  I didn't know about this.

File: tests/unit-tests/utils.spec.ts (hunk 7 above)
Comment (pstaabp) [reply tests/unit-tests/utils.spec.ts:5]: Got it.

File: t/mojolicious/003_users.t (hunk 6 above)
Comment (pstaabp) [reply t/mojolicious/003_users.t:95]: Actually, don't we need a grep here, that is:
```
my $admin_user = (grep { $_->{username} eq 'admin' } @all_users)[0];
```

**Comment by pstaabp** [comment]
I just did the recommended fixes.  A couple of notes:
- I created an inflator class to handle to the boolean--great idea--now you can just have a column type called 'json_bool' that just handles it. 
- changed all roles to lowercase.
- suggested code cleanup including removing `List::MoreUtils` and `firstval`

I didn't look into the UI.  I think that @drdrew42 session handling PR should fix this if we're okay with skipping fixing this.

**Comment by drdrew42** [comment]
Once this is approved + merged, I have a fix for maintaining session duration in state, as well as route guarding and preserving the target URL until after login is completed.

**Comment by drdrew42** [comment]
there is now a preview of my changes at pstaabp/webwork3#29

**Comment by drdrew42** [comment]
I think changing capitalization of roles has more instances that need to be dealt with... e.g. logging in with lisa displays no courses -- they're in the store, but have roles in all caps still.

**Comment by drgrice1** [comment]
> I just did the recommended fixes. A couple of notes:
> 
>     * I created an inflator class to handle to the boolean--great idea--now you can just have a column type called 'json_bool' that just handles it.
This isn't working at all.  I am a little surprised that sqlite works with this, but this completely fails with mysql and postgres.  I suggest that in the future you test all database changes with both.  Just run `./bin/setup_db.pl` with `webwork3.yml` set for either mysql or postgres to see what I mean.  The problem is that `json_bool` is not a valid database type.  You note in the `lib/DBIx/Class/InflateColumn/JSONBool.pm` file that it is a simplified version of `DBIx::Class::InflateColumn::DateTime`.  Note that the reason that package works is because `date` and `datetime` are valid database types.

In testing this I found another issue.  Data type `text` can not be used in a unique constraint.  You did this in `lib/DB/Schema/Result/Role.pm`.  The data type of the `role_name` column will need to be switched to `varchar`.

I put a pull request into your branch for this pull request that fixes these database issues.
> 
>     * changed all roles to lowercase.
> 
>     * suggested code cleanup including removing `List::MoreUtils` and `firstval`
This is certainly not finished.  You changed one instance (the one I specifically commented on).  The point is to entirely remove `firstval` from the code and thus the dependence on `List::MoreUtils`.  Also, my comment was a note for the future.  So don't do this in this pull request.  That should be done in a future pull request that is dedicated to doing only that.
> 
> 
> I didn't look into the UI. I think that @drdrew42 session handling PR should fix this if we're okay with skipping fixing this.
I am certainly not okay with merging a pull request that makes the UI completely dis-functional.

**Comment by drdrew42** [comment]
So should we keep pstaabp/webwork3#29 open for review before we merge it into this branch?

**Comment by drdrew42** [comment]
I will also switch over to postgres so that I'm testing with a more production-like environment.

**Comment by drgrice1** [comment]
> So should we keep [pstaabp#29](https://github.com/pstaabp/webwork3/pull/29) open for review before we merge it into this branch?

Is that the "session handling PR" that @pstaabp referred to?  I didn't realize that he was referring to a pull request that adds to this one.  If that fixes the UI issue, then once that is merged we can merge this one.  Yeah, let me look at that pull request before it is merged into this one.

**Comment by drdrew42** [comment]
Yes, I rebased my 'session handling' changes off of the branch for this PR. (And I opened the PR to @pstaabp's fork so that it could be reviewed without all the other changes already seen here.)

**Comment by drdrew42** [comment]
I also kept my changes separate in case we decided to squash the commits here.

**Comment by drdrew42** [comment]
`build_db.pl` needs changes to run, in addition to pstaabp/webwork3#30:
- insert new line at L110 `my $student_role = $role_rs->find({ role_name => 'student' });`
- edit L117 (now L118) `$course->add_to_users($stud_info, { role_id => $student_role->role_id });`

**Comment by drgrice1** [comment]
@drdrew42: pstaabp#29 looks good for now and fixes the UI issue.

I added the lines you suggested to my pull request to this branch.  Those are certainly needed to use the build_db.pl script with mysql or postgres, otherwise there is a failed foreign key constraint issue.  I hadn't tested using the build_db.pl script with either of those databases.  I just had run setup_db.pl and saw the other issues.  Thanks for pointing that out.

**Comment by pstaabp** [comment]
I'll be honest, I've only been running test on sqlite.  I'll switch to others before committing code.  Switching over to mariadb/mysql I'm getting test failing.  At least some are on arrays in which the order is flipped. @drgrice1 are all the tests running for you on mysql and Postgres?

**Comment by drgrice1** [comment]
@pstaabp:  First, I should point out that when I say mysql I really mean mariadb.  I forget that I am actually running mariadb, largely because other than the ubuntu package name mysql and mariadb are the same.  Both use the same command line command name and same service name.

I am seeing that many of the unit tests are failing with both mariadb and postgres.  I am starting to think that using sqlite is a bad idea.  It is to lenient and allows with many things that more proper databases do not.  I think that we are going to need to rework the unit tests to test with both mariadb and postgres.

**Comment by pstaabp** [comment]
I'm running mariadb right now and all but one of the failing tests is on an array and sorting them before testing fixes them.  I am having an issue with course_user and getting a fail on a addCourseUser.  It seems to be an issue with the many-to-many bridge table in mariadb. 

If I can get these fixed, I'll make sure I test in Postgres too.

**Comment by drdrew42** [comment]
### Test Revisions Needed
- [ ] add quote_names => to schema connect
- [ ] handle sorting in tests (TBD: extend sort support to API queries)
- [ ] when using add_to_$rel on a many-to-many relation, use second argument to add params to merge table row

**Comment by pstaabp** [comment]
Made a number of changes to get the test to run on mysql.  @drdrew42 said he will try under Postgres (I'll try later as well). 

Main fixes are:
- sorting arrays that come out of the database, so comparisons are made on sorted arrays.
- for UserProblem on the status field, the data_type was set right.  I have a numeric of size 10,5 (5 decimal places). Tests need to compare to numbers as strings to 5 decimal places.

**Comment by pstaabp** [comment]
The tests are failing under sqlite now.  Will find a better solution to these.

**Comment by drdrew42** [comment]
tests 007, 010, and 012 in t/db/ still need `quote_names => 1` (do not trust search-and-replace without checking that every expected file has been touched?)

**Comment by drgrice1** [comment]
At this point the tests are failing for me with all databases.  I tested with mariadb, postgres, and sqlite.

**Comment by pstaabp** [comment]
> tests 007, 010, and 012 in t/db/ still need `quote_names => 1` (do not trust search-and-replace without checking that every expected file has been touched?)

Will update 007, 010 and 012.  

> At this point the tests are failing for me with all databases. I tested with mariadb, postgres, and sqlite.

I may need to rethink the tests (especially those involving comparing two decimal numbers).

**Comment by pstaabp** [comment]
I'm seeing all tests pass on sqlite, mysql/mariadb and Postgres.

File: src/common/models/users.ts [hunk 12]
```diff
@@ -187,7 +183,7 @@ export class CourseUser extends Model {
 	private _first_name = '';
 	private _last_name = '';
 	private _student_id = '';
-	private _role = UserRole.unknown;
+	private _role = 'UNKNOWN';
```
Comment (drgrice1) [inline src/common/models/users.ts:186]: UNKNOWN should be lower case here.

File: src/common/models/users.ts [hunk 13]
```diff
@@ -266,7 +258,7 @@ export class CourseUser extends Model {
 	isValid(): boolean {
 		return isNonNegInt(this.user_id) && isNonNegInt(this.course_user_id) &&
 			isNonNegInt(this.course_id) && isValidUsername(this.username) &&
-			this.role !== UserRole.unknown && (this.email === '' || isValidEmail(this.email));
+			this.role !== 'UNKNOWN' && (this.email === '' || isValidEmail(this.email));
```
Comment (drgrice1) [inline src/common/models/users.ts:261]: UNKNOWN should be lower case here.

File: src/common/models/users.ts [hunk 14]
```diff
@@ -276,7 +268,7 @@ export class CourseUser extends Model {
 			user_id: isNonNegInt(this.user_id) || 'The user_id must be a non negative integer.',
 			username: isValidUsername(this.username) || 'The username must be valid.',
 			email: (this.email === '' || isValidEmail(this.email)) || 'The email must be valid',
-			role: this.role !== UserRole.unknown || 'The role is not valid.',
+			role: this.role !== 'UNKNOWN' || 'The role is not valid.',
```
Comment (drgrice1) [inline src/common/models/users.ts:271]: UNKNOWN should be lower case here.

**Review by drgrice1 — COMMENTED** [review]
There are a few minor changes to the role casing that are still needed.

Line 198 of src/components/instructor/ClassListManagerComponents/AddUsersFromFile.vue is another instance of where UNKNOWN should be lower case.  Furthermore there is a mispelled UNKOWN on that line.

Line 143 of tests/unit-tests/course_users.spec.ts has another instance where UNKNOWN should be lower case.

I am not sure that I agree that there should be a `course_admin` role.  We should talk about this.

**Comment by drgrice1** [comment]
Once the casing instances are fixed we can merge this.

**Comment by pstaabp** [comment]
Just fixed these.  Noticed that the problem set types are also upper case.  I will change those in a separate PR. 

About the 'course_admin' role, I'm not wedded to it, but just building the infrastructure and if we have permissions set up right we should be able to change roles and what permissions they have in the conf file.

**Comment by drgrice1** [comment]
Once the tests are complete I will merge this.

@drdrew42:  Are you okay with a squash merge?  Doing so will remove our commits, however the commit message in the new squashed commit will list us as co-authors.  Github automatically adds
```
Co-authored-by: K. Andrew Parker <kparker@citytech.cuny.edu>
Co-authored-by: Glenn Rice <grice1@missouriwestern.edu>
```
to the end.

@pstaabp:  I did notice the uppercase problem set types.

**Comment by drgrice1** [comment]
> Line 143 of tests/unit-tests/course_users.spec.ts has another instance where UNKNOWN should be lower case.

This line didn't get changed.  That is why the tests failed.

**Comment by pstaabp** [comment]
I'm fine with a squash merge.  Uggh.  Another fail!!

**Comment by pstaabp** [comment]
And trying to be more focused with PRs and  I have figured out the rebasing--the latest two PRs I have have only the commits on those branches, so moving ahead, these should be cleaner.

**Review by drdrew42 — APPROVED** [review]
Let's squash

**Comment by drgrice1** [comment]
Okay.  It is in.
